libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1763,31 +1800,95 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
 {
 	Type const* leftType = type(_operation.leftExpression());
 	Type const* rightType = type(_operation.rightExpression());
-	TypeResult result = leftType->binaryOperatorResult(_operation.getOperator(), rightType);
-	Type const* commonType = result.get();
-	if (!commonType)
+
+	// Check if the operator is built-in or user-defined.
+	TypeResult builtinResult = leftType->binaryOperatorResult(_operation.getOperator(), rightType);
+	set<FunctionDefinition const*, ASTNode::CompareByID> matchingDefinitions = leftType->operatorDefinitions(
+		_operation.getOperator(),
+		*currentDefinitionScope(),
```

Comment:
Which scope is this exactly? Isn't this the scope that contains the operation itself? Isn't that the wrong scope for the lookup?

Reply:
To be perfectly honest, I still don't fully understand how our scopes work but this is the way @chriseth did it originally and I did not find any test case that would break it so I was assuming it's correct.

On the surface, it does seem to make sense. It's the scope where the operator is used so you're only interested in operators visible there.